clang/test/Sema/attr-nonblocking-sema.cpp

```diff
@@ -0,0 +1,141 @@
+  fp_nonblocking = nonallocating; // expected-warning {{attribute 'nonblocking' should not be added via type conversion}}
+
+  // Adding/spoofing nonallocating is unsafe.
+  void (*fp_nonallocating)() [[clang::nonallocating]];
+  fp_nonallocating = nullptr;
+  fp_nonallocating = nonallocating;
+  fp_nonallocating = nonblocking; // no warning because nonblocking includes nonallocating fp_nonallocating = unannotated;
+  fp_nonallocating = unannotated; // expected-warning {{attribute 'nonallocating' should not be added via type conversion}}
+}
+
+#ifdef __cplusplus
+// There was a bug: noexcept and nonblocking could be individually removed in conversion, but not both  
+void type_conversions_2()
+{
+  auto receives_fp = [](void (*fp)()) {
+  };
+  
+  auto ne = +[]() noexcept {};
+  auto nl = +[]() [[clang::nonblocking]] {};
+  auto nl_ne = +[]() noexcept [[clang::nonblocking]] {};
+  
+  receives_fp(ne);
+  receives_fp(nl);
+  receives_fp(nl_ne);
+}
+#endif
+
+// --- VIRTUAL METHODS ---
+// Attributes propagate to overridden methods, so no diagnostics except for conflicts.
+// Check this in the syntax tests too.
+#ifdef __cplusplus
+struct Base {
+  virtual void f1();
+  virtual void nonblocking() noexcept [[clang::nonblocking]];
+  virtual void nonallocating() noexcept [[clang::nonallocating]];
+  virtual void f2() [[clang::nonallocating]]; // expected-note {{previous declaration is here}}
+};
+
+struct Derived : public Base {
+  void f1() [[clang::nonblocking]] override;
+  void nonblocking() noexcept override;
+  void nonallocating() noexcept override;
+  void f2() [[clang::allocating]] override; // expected-warning {{effects conflict when merging declarations; kept 'allocating', discarded 'nonallocating'}}
+};
+#endif // __cplusplus
+
+// --- REDECLARATIONS ---
+
+void f2();
+void f2() [[clang::nonblocking]]; // expected-note {{previous declaration is here}}
+void f2(); // expected-warning {{attribute 'nonblocking' on function does not match previous declaration}}
+// Note: we verify that the attribute is actually seen during the constraints tests.
+
+void f3() [[clang::blocking]]; // expected-note {{previous declaration is here}}
+void f3() [[clang::nonblocking]]; // expected-warning {{effects conflict when merging declarations; kept 'blocking', discarded 'nonblocking'}}
+
+// --- OVERLOADS ---
+#ifdef __cplusplus
+struct S {
+  void foo(); // expected-note {{previous declaration is here}}
+  void foo(); // expected-error {{class member cannot be redeclared}}
+};
+#endif // __cplusplus
+
+// --- COMPUTED NONBLOCKING ---
+void f4() [[clang::nonblocking(__builtin_memset)]] {} // expected-error {{nonblocking attribute requires an integer constant}}
+
+#ifdef __cplusplus
+// Unexpanded parameter pack
+template <bool ...val>
+void f5() [[clang::nonblocking(val /* NO ... here */)]] {} // expected-error {{expression contains unexpanded parameter pack 'val'}}
+
```

Comment:
done, at least partially. This gets diagnosed at the declaration of `func()`. The diagnostic is slightly odd in that it says `nonblocking(expr)` and `blocking` are incompatible, without being able to say what `expr` is.

Fixing that might require that we stop trying to generate diagnostics from a templated declaration, and instead do a different variation of the check after template instantiation. (?)

Reply:
Yeah, generally speaking you should check whether the expression is dependent, and if it is, defer the checking until template instantiation happens (as template instantiation eventually calls `ProcessDeclAttribute()` again with the non-dependent value).

Reply:
OK, this needs a bit of refactoring to completely separate attribute parsing from validation.

Reply:
It's not quite implemented the way you suggested because the attributes are represented in the type; now when TreeTransform transforms the `FunctionProtoType`'s effects, any conditions are transformed, and a newly-extracted validation method is called.  I didn't quite know what to expect!  But this seems unsurprising:

```
/xwork/attr-nolock-wip.cpp:7:3: error: no matching function for call to 'ambiguous'
    7 |   ambiguous<true>();
      |   ^~~~~~~~~~~~~~~
/xwork/attr-nolock-wip.cpp:4:6: note: candidate template ignored: substitution failure [with B = true]: 'blocking' and 'nonblocking' attributes are not compatible
    4 | void ambiguous() [[clang::nonblocking(B)]] [[clang::blocking]];
      | ~~~~ ^
1 error generated.
```

Reply:
And a call to `ambiguous<false>()` also fails, because currently it is an error to specify the same attribute twice, in the simplest case:

```
void foo() [[clang::nonblocking]] [[clang::nonblocking]];
```

I believe this is consistent with other attributes (?) but maybe it should be relaxed for the benefit of this case with a template?

Reply:
I discovered that having two instances of `[[nodiscard]]` on a declaration is ignored as harmless, so I decided to mimic that, so that things behave more sensibly when one of the two attributes was expanded from e.g. `nonblocking(B)`.

Reply:
That sounds reasonable and also matches how e.g. other attributes and even `const` is handled in some cases that involve templates iirc.

Reply:
Yeah, that's the way I'd expect it to work as well.